### DISCUSSION TITLE
[@mantine/modals] RFC: Make context modals fully typesafe

### DISCUSSION THREAD

**Discussion by Meierschlumpf** [mantinedev/mantine#3161] [Feature requests]
Hey it's me again!

I was able to make modals fully typesafe with some changes on the typescript definitions and the usage of declare module.
Maybe you know about @tanstack/router which also uses that to make there router fully typesafe.

```typescript
const ModalA = ({id, innerProps, context}: ContextModalProps<{id: number}>) => {
  return <Text>Modal A</Text>
}

const ModalB = ({id, innerProps, context}: ContextModalProps<{todo: { id: number, title: string, isDone: boolean }}>) => {
  return <Text>Modal B</Text>
}


const modals = createModals({
  modalA: ModalA,
  modalB: ModalB,
});

declare module '@mantine/modals/lib/context' {
  interface RegisterModals {
    modals: typeof modals;
  }
}

openContextModal({
  modal: 'modalA',
  innerProps: {
    id: 5
  },
});
```

![image](https://user-images.githubusercontent.com/63781622/206856314-fa33ebc0-d6d6-4978-8b84-65c08bb5860a.png)

The same would of course be possible with closeModal and all the other things. I just need your opinion and go for that I can start implementing it for mantine.

**Comment by rtivital** [comment]
I'm fine with it. It should be implemented the same way as `theme.other` and `theme.colors` types overrides.

**Reply by Meierschlumpf** [reply]
Hey @rtivital,
I noticed, that my pr has been dumped cause of a new version. (The branch had been deleted)
So I asked myself, if I should add this feature to the v6 branch? Would make sense in my opinion, because it has a small breaking change within typescript.

**Reply by rtivital** [reply]
Yes, feel free to send a PR to v6 branch after 5.10.2 release is out
